ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Closure.cshtml.cs

```diff
@@ -148,11 +152,13 @@ private async Task<List<string>> ValidateCloseConcern(long caseUrn)
 			var financialPlanModels = (await _financialPlanModelService.GetFinancialPlansModelByCaseUrn(caseUrn, User.Identity.Name)).ToList();
 			var ntiUnderConsiderationModels = (await _ntiUnderConsiderationModelService.GetNtiUnderConsiderationsForCase(caseUrn)).ToList();
 			var ntiWarningLetterModels = (await _ntiWarningLetterModelService.GetNtiWarningLettersForCase(caseUrn)).ToList();
+			var ntiModels = (await _ntiModelService.GetNtisForCaseAsync(caseUrn)).ToList();
 
 			var hasOpenSRMAs = srmaModels.Any(srma => srma.ClosedAt == null) == true;
 			var hasOpenFinancialPlans = financialPlanModels.Any(fp => fp.ClosedAt == null) == true;
 			var hasOpenNTIUnderConsiderations = ntiUnderConsiderationModels.ToList().Any(uc => uc.ClosedAt == null) == true;
-			var hasOpenNTIWarningLetters = ntiWarningLetterModels.ToList().Any(uc => uc.ClosedAt == null) == true;
+			var hasOpenNTIWarningLetters = ntiWarningLetterModels.ToList().Any(wl => wl.ClosedAt == null) == true;
+			var hasOpenNTIs = ntiModels.ToList().Any(nti => nti.ClosedAt == null) == true;
 
 

```

Comment:
An alternative to writing an extra 4 lines of code per validation rule would be to either inject separate rules (/strategies) and iterate them all (inject them as an array and let IoC give them to the method).
Or define them within this method differently, for example....

```
var validationRules = new (Func<Task<bool>> predicate, System.Action action)[]
			{
				new (() => Task.FromResult(numberOfOpenConcerns>0), () => errorMessages.Add("Resolve Concerns")),
				new (() => Task.FromResult(hasOpenSRMAs), () => errorMessages.Add("Resolve Concerns")),
				new (() => Task.FromResult(hasOpenFinancialPlans), () => errorMessages.Add("Resolve Financial Plan")),
				new (() => Task.FromResult(hasOpenNTIUnderConsiderations), () => errorMessages.Add("Resolve NTI Under Consideration")),
				new (() => Task.FromResult(hasOpenNTIWarningLetters), () => errorMessages.Add("Resolve NTI Warning Letter")),
				new (() => Task.FromResult(hasOpenNTIs), () => errorMessages.Add("Resolve NTI Warning Letter")),
				new (async () => await IsCaseAlreadyClosed(User.Identity.Name, caseUrn), () =>  errorMessages.Add("This case is already closed."))

			};

			foreach (var rule in validationRules)
			{
				var ruleViolated = await rule.predicate();
				if (ruleViolated)
				{
					rule.action();
				}
			}
```

Reply:
Also, the validation rules are likely to be better if they're in the models, rather than the page behind.
This would help us build up a smarter domain model (rather than an anaemic one).

In my example above you could probably get rid of the intermediary variables (eg hasOpenSRMAs) and instead ask each of the models to validate or define validators in the domain and each each validator to validate models (if that makes sense).

Basically moving the logic out of the page and into the models